### PR TITLE
fix(developer): make native compilation of kmcmplib under Linux possible

### LINKED ISSUE #11772: bug(developer): kmcmplib fails native compilation under Linux
### Describe the bug

I have installed Ubuntu-20.02 under WSL2 on Windows 10 using the instructions for [Ubuntu](https://github.com/keymanapp/keyman/blob/17d40ff96d770cd95e5d8c6f862566334772b134/docs/build/linux-ubuntu.md),  I included the [keyman-alpha ppa](https://launchpad.net/~keymanapp/+archive/ubuntu/keyman-alpha).  I have installed `emscripten` 3.1.44 and `node`  18.20.3.

The `wsl -v` details are:
```
WSL version: 2.1.5.0
Kernel version: 5.15.146.1-2
WSLg version: 1.0.60
MSRDC version: 1.2.5105
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.4529
```

I then attempted to test `kmcmplib` in `$HOME/keyman/developer/src/kmcmplib`.  For wasm (` ./build.sh test:wasm`), all is well, but the native build fails (`./build.sh test:arch`).

The trace:

```
[developer/src/kmcmplib] build.sh parameters: <test:arch>
[developer/src/kmcmplib] Automatically running following required actions with missing outputs:
[developer/src/kmcmplib] * build:arch
[developer/src/kmcmplib] ## build:arch starting...
[common/include] dependency build, started by developer/src/kmcmplib
[common/include] build.sh parameters: <configure build --deps --builder-dep-parent developer/src/kmcmplib>
[common/include] skipping configure, up-to-date
[common/include] skipping build, up-to-date
[1/13] Compiling C++ object src/libkmcmplib.a.p/NamedCodeConstants.cpp.o
FAILED: src/libkmcmplib.a.p/NamedCodeConstants.cpp.o
c++ -Isrc/libkmcmplib.a.p -Isrc -I../../../src -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -fPIC -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -Wctor-dtor-privacy -Wdouble-promotion -Wendif-labels -Wno-unknown-pragmas -Wno-missing-field-initializers -Wnon-virtual-dtor -Wshadow -Wall -Wextra -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ src/libkmcmplib.a.p/NamedCodeConstants.cpp.o -MF src/libkmcmplib.a.p/NamedCodeConstants.cpp.o.d -o src/libkmcmplib.a.p/NamedCodeConstants.cpp.o -c ../../../src/NamedCodeConstants.cpp
../../../src/NamedCodeConstants.cpp:35:15: error: expected initializer before ‘sort_entries’
   35 |   int __cdecl sort_entries(const void *elem1, const void *elem2)
      |               ^~~~~~~~~~~~
../../../src/NamedCodeConstants.cpp: In member function ‘void kmcmp::NamedCodeConstants::reindex()’:
../../../src/NamedCodeConstants.cpp:172:48: error: ‘sort_entries’ was not declared in this scope
  172 |     qsort(entries, nEntries, sizeof(NCCENTRY), sort_entries);
      |                                                ^~~~~~~~~~~~
../../../src/NamedCodeConstants.cpp: In function ‘int IsHangulSyllable(const KMX_WCHAR*, int*)’:
../../../src/NamedCodeConstants.cpp:271:45: error: comparison of integer expressions of different signedness: ‘wint_t’ {aka ‘unsigned int’} and ‘int’ [-Werror=sign-compare]
  271 |     int isDoubled = towupper(*(codename+1)) == ch;
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
cc1plus: all warnings being treated as errors
[2/13] Compiling C++ object src/libkmcmplib.a.p/kmx_u16.cpp.o
FAILED: src/libkmcmplib.a.p/kmx_u16.cpp.o
c++ -Isrc/libkmcmplib.a.p -Isrc -I../../../src -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -fPIC -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -Wctor-dtor-privacy -Wdouble-promotion -Wendif-labels -Wno-unknown-pragmas -Wno-missing-field-initializers -Wnon-virtual-dtor -Wshadow -Wall -Wextra -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ src/libkmcmplib.a.p/kmx_u16.cpp.o -MF src/libkmcmplib.a.p/kmx_u16.cpp.o.d -o src/libkmcmplib.a.p/kmx_u16.cpp.o -c ../../../src/kmx_u16.cpp
In file included from ../../../src/kmx_u16.cpp:3:
../../../include/kmcompx.h:33:24: error: ‘size_t’ has not been declared
   33 | template < typename T, size_t N >
      |                        ^~~~~~
../../../include/kmcompx.h:34:1: error: ‘size_t’ does not name a type
   34 | size_t _countof( T ( & /*arr*/ )[ N ] )
      | ^~~~~~
../../../include/kmcompx.h:32:1: note: ‘size_t’ is defined in header ‘<cstddef>’; did you forget to ‘#include <cstddef>’?
   31 | #include <type_traits>
  +++ |+#include <cstddef>
   32 |
../../../src/kmx_u16.cpp: In function ‘std::string toHex(int)’:
../../../src/kmx_u16.cpp:99:2: error: ‘reverse’ was not declared in this scope
   99 |  reverse(s.begin(), s.end());
      |  ^~~~~~~
../../../src/kmx_u16.cpp: In function ‘KMX_CHAR* strrchr_slash(KMX_CHAR*)’:
../../../src/kmx_u16.cpp:128:8: error: ‘strrchr’ was not declared in this scope
  128 |   cp = strrchr(Name, '\\');
      |        ^~~~~~~
../../../src/kmx_u16.cpp:12:1: note: ‘strrchr’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
   11 | #include <stdarg.h>
  +++ |+#include <cstring>
   12 |
../../../src/kmx_u16.cpp: In function ‘double u16tof(KMX_WCHAR*)’:
../../../src/kmx_u16.cpp:279:14: error: comparison of unsigned expression < 0 is always false [-Werror=type-limits]
  279 |  if (pos_dot < 0)
      |      ~~~~~~~~^~~
cc1plus: all warnings being treated as errors
[3/13] Compiling C++ object tests/api-test.p/util_callbacks.cpp.o
FAILED: tests/api-test.p/util_callbacks.cpp.o
c++ -Itests/api-test.p -Itests -I../../../tests -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ tests/api-test.p/util_callbacks.cpp.o -MF tests/api-test.p/util_callbacks.cpp.o.d -o tests/api-test.p/util_callbacks.cpp.o -c ../../../tests/util_callbacks.cpp
../../../tests/util_callbacks.cpp: In function ‘int msgproc(int, uint32_t, const char*, void*)’:
../../../tests/util_callbacks.cpp:19:10: error: '0' flag ignored with precision and ‘%x’ gnu_printf format [-Werror=format=]
   19 |   printf("line %d  %s %04.4x:  %s\n", line, t, (unsigned int)dwMsgCode, szText);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../tests/util_callbacks.cpp: In function ‘bool loadfileProc(const char*, const char*, void*, int*, void*)’:
../../../tests/util_callbacks.cpp:52:34: error: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Werror=sign-compare]
   52 |     if(fread(data, 1, *size, fp) != *size) {
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
cc1plus: all warnings being treated as errors
[4/13] Compiling C++ object tests/kmcompxtest.p/util_callbacks.cpp.o
FAILED: tests/kmcompxtest.p/util_callbacks.cpp.o
c++ -Itests/kmcompxtest.p -Itests -I../../../tests -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ tests/kmcompxtest.p/util_callbacks.cpp.o -MF tests/kmcompxtest.p/util_callbacks.cpp.o.d -o tests/kmcompxtest.p/util_callbacks.cpp.o -c ../../../tests/util_callbacks.cpp
../../../tests/util_callbacks.cpp: In function ‘int msgproc(int, uint32_t, const char*, void*)’:
../../../tests/util_callbacks.cpp:19:10: error: '0' flag ignored with precision and ‘%x’ gnu_printf format [-Werror=format=]
   19 |   printf("line %d  %s %04.4x:  %s\n", line, t, (unsigned int)dwMsgCode, szText);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../tests/util_callbacks.cpp: In function ‘bool loadfileProc(const char*, const char*, void*, int*, void*)’:
../../../tests/util_callbacks.cpp:52:34: error: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Werror=sign-compare]
   52 |     if(fread(data, 1, *size, fp) != *size) {
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
cc1plus: all warnings being treated as errors
[5/13] Compiling C++ object tests/kmcompxtest.p/kmcompxtest.cpp.o
FAILED: tests/kmcompxtest.p/kmcompxtest.cpp.o
c++ -Itests/kmcompxtest.p -Itests -I../../../tests -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ tests/kmcompxtest.p/kmcompxtest.cpp.o -MF tests/kmcompxtest.p/kmcompxtest.cpp.o.d -o tests/kmcompxtest.p/kmcompxtest.cpp.o -c ../../../tests/kmcompxtest.cpp
../../../tests/kmcompxtest.cpp: In function ‘int main(int, char**)’:
../../../tests/kmcompxtest.cpp:67:22: error: ‘strrchr’ was not declared in this scope
   67 |     char* testname = strrchr( (char*) kmn_file, '/') + 1;
      |                      ^~~~~~~
../../../tests/kmcompxtest.cpp:22:1: note: ‘strrchr’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
   21 | #include <unistd.h>
  +++ |+#include <cstring>
   22 | #endif
../../../tests/kmcompxtest.cpp:68:8: error: ‘strncmp’ was not declared in this scope
   68 |     if(strncmp(testname, pfirst5, 5) == 0){
      |        ^~~~~~~
../../../tests/kmcompxtest.cpp:68:8: note: ‘strncmp’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
../../../tests/kmcompxtest.cpp:88:24: error: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
   88 |     if (result.kmxSize != sz2) return __LINE__;                //  exit code: size of kmx-file in build differs from size of kmx-file in source folder
      |         ~~~~~~~~~~~~~~~^~~~~~
../../../tests/kmcompxtest.cpp:92:12: error: ‘memcmp’ was not declared in this scope
   92 |     return memcmp(result.kmx, buf2, result.kmxSize) ? __LINE__ : 0;  // exit code:  when contents of kmx-file in build differs from contents of kmx-file in source folder
      |            ^~~~~~
../../../tests/kmcompxtest.cpp:92:12: note: ‘memcmp’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
../../../tests/kmcompxtest.cpp:97:22: error: ‘strrchr’ was not declared in this scope
   97 |     char* testname = strrchr( (char*) kmn_file, '/') + 1;
      |                      ^~~~~~~
../../../tests/kmcompxtest.cpp:97:22: note: ‘strrchr’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
../../../tests/kmcompxtest.cpp:100:10: error: ‘strncmp’ was not declared in this scope
  100 |     if ((strncmp(testname, pfirst5, 5) == 0) &&   (testname[9] == '_')) {
      |          ^~~~~~~
../../../tests/kmcompxtest.cpp:100:10: note: ‘strncmp’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
../../../tests/kmcompxtest.cpp:101:21: error: ‘strchr’ was not declared in this scope
  101 |       char* ErrNr = strchr(testname, '_') +1 ;
      |                     ^~~~~~
../../../tests/kmcompxtest.cpp:101:21: note: ‘strchr’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
../../../tests/kmcompxtest.cpp:105:25: error: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<int>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
  105 |       for (int i = 0; i < error_vec.size() ; i++) {
      |                       ~~^~~~~~~~~~~~~~~~~~
../../../tests/kmcompxtest.cpp: In function ‘bool isDesktopKeyboard(FILE*)’:
../../../tests/kmcompxtest.cpp:159:20: error: comparison of integer expressions of different signedness: ‘int’ and ‘KMX_DWORD’ {aka ‘unsigned int’} [-Werror=sign-compare]
  159 |   for(int i = 0; i < fk.cxStoreArray; i++, s++) {
      |                  ~~^~~~~~~~~~~~~~~~~
../../../tests/kmcompxtest.cpp: In function ‘int main(int, char**)’:
../../../tests/kmcompxtest.cpp:91:10: error: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Werror=unused-result]
   91 |     fread(buf2, 1, result.kmxSize, fp2);
      |     ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
[6/13] Compiling C++ object src/libkmcmplib.a.p/CompileKeyboardBuffer.cpp.o
FAILED: src/libkmcmplib.a.p/CompileKeyboardBuffer.cpp.o
c++ -Isrc/libkmcmplib.a.p -Isrc -I../../../src -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -fPIC -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -Wctor-dtor-privacy -Wdouble-promotion -Wendif-labels -Wno-unknown-pragmas -Wno-missing-field-initializers -Wnon-virtual-dtor -Wshadow -Wall -Wextra -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ src/libkmcmplib.a.p/CompileKeyboardBuffer.cpp.o -MF src/libkmcmplib.a.p/CompileKeyboardBuffer.cpp.o.d -o src/libkmcmplib.a.p/CompileKeyboardBuffer.cpp.o -c ../../../src/CompileKeyboardBuffer.cpp
../../../src/CompileKeyboardBuffer.cpp: In function ‘void kmcmp::CopyExtraData(PFILE_KEYBOARD)’:
../../../src/CompileKeyboardBuffer.cpp:177:22: error: comparison of integer expressions of different signedness: ‘int’ and ‘KMX_DWORD’ {aka ‘unsigned int’} [-Werror=sign-compare]
  177 |     for(int i = 0; i < fk->cxStoreArray; i++, store++) {
      |                    ~~^~~~~~~~~~~~~~~~~~
../../../src/CompileKeyboardBuffer.cpp:191:22: error: comparison of integer expressions of different signedness: ‘int’ and ‘KMX_DWORD’ {aka ‘unsigned int’} [-Werror=sign-compare]
  191 |     for(int i = 0; i < fk->cxGroupArray; i++, group++) {
      |                    ~~^~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
[7/13] Compiling C++ object src/libkmcmplib.a.p/UnreachableRules.cpp.o
FAILED: src/libkmcmplib.a.p/UnreachableRules.cpp.o
c++ -Isrc/libkmcmplib.a.p -Isrc -I../../../src -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -fPIC -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -Wctor-dtor-privacy -Wdouble-promotion -Wendif-labels -Wno-unknown-pragmas -Wno-missing-field-initializers -Wnon-virtual-dtor -Wshadow -Wall -Wextra -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ src/libkmcmplib.a.p/UnreachableRules.cpp.o -MF src/libkmcmplib.a.p/UnreachableRules.cpp.o.d -o src/libkmcmplib.a.p/UnreachableRules.cpp.o -c ../../../src/UnreachableRules.cpp
../../../src/UnreachableRules.cpp: In function ‘std::wstring kmcmp::MakeHashKeyFromFileKey(PFILE_KEY)’:
../../../src/UnreachableRules.cpp:21:63: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   21 |       std::wstring Context_ws = u16fmt((const PKMX_WCHAR) kp->dpContext);
      |                                                               ^~~~~~~~~
cc1plus: all warnings being treated as errors
[8/13] Compiling C++ object src/libkmcmplib.a.p/Compiler.cpp.o
FAILED: src/libkmcmplib.a.p/Compiler.cpp.o
c++ -Isrc/libkmcmplib.a.p -Isrc -I../../../src -Iinclude -I../../../include -I../../../../common/include -I../../../../../../common/include -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=c++14 -O3 -g -fPIC -DU_STATIC_IMPLEMENTATION -DU_ATTRIBUTE_DEPRECATED= -DUCONFIG_USE_LOCAL=1 -DU_NOEXCEPT= -DKMN_KBP_EXPORTING -Wctor-dtor-privacy -Wdouble-promotion -Wendif-labels -Wno-unknown-pragmas -Wno-missing-field-initializers -Wnon-virtual-dtor -Wshadow -Wall -Wextra -fvisibility=hidden -fvisibility-inlines-hidden -MD -MQ src/libkmcmplib.a.p/Compiler.cpp.o -MF src/libkmcmplib.a.p/Compiler.cpp.o.d -o src/libkmcmplib.a.p/Compiler.cpp.o -c ../../../src/Compiler.cpp
../../../src/Compiler.cpp: In function ‘bool resizeKeyArray(PFILE_GROUP, int)’:
../../../src/Compiler.cpp:809:45: error: comparison of integer expressions of different signedness: ‘KMX_DWORD’ {aka ‘unsigned int’} and ‘int’ [-Werror=sign-compare]
  809 |   if((gp->cxKeyArray + increment - 1) % 100 < increment) {
      |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../../src/Compiler.cpp: In function ‘KMX_DWORD ProcessSystemStore(PFILE_KEYBOARD, KMX_DWORD, PFILE_STORE)’:
../../../src/Compiler.cpp:1052:70: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
 1052 |       KMX_WCHAR *pp2 = (KMX_WCHAR*) u16rchr_slash((const PKMX_WCHAR) p);
      |                                                                      ^
../../../src/Compiler.cpp: In function ‘int GetCompileTargetsFromTargetsStore(const KMX_WCHAR*)’:
../../../src/Compiler.cpp:1194:14: error: declaration of ‘p’ shadows a previous local [-Werror=shadow]
 1194 |     for(auto p: KMXKeymanTargets) {
      |              ^
../../../src/Compiler.cpp:1186:8: note: shadowed declaration is here
 1186 |   auto p = new KMX_WCHAR[u16len(store)+1];
      |        ^
../../../src/Compiler.cpp:1197:14: error: declaration of ‘p’ shadows a previous local [-Werror=shadow]
 1197 |     for(auto p: KMWKeymanTargets) {
      |              ^
../../../src/Compiler.cpp:1186:8: note: shadowed declaration is here
 1186 |   auto p = new KMX_WCHAR[u16len(store)+1];
      |        ^
../../../src/Compiler.cpp: In function ‘KMX_DWORD GetRHS(PFILE_KEYBOARD, PKMX_WCHAR, PKMX_WCHAR, int, int, int)’:
../../../src/Compiler.cpp:3286:39: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
 3286 |   p = (const PKMX_WCHAR) u16chr(p, '>');
      |                                       ^
../../../src/Compiler.cpp: In function ‘bool UTF16TempFromUTF8(KMX_BYTE*, int, KMX_BYTE**, int*)’:
../../../src/Compiler.cpp:3519:28: error: catching polymorphic type ‘class std::range_error’ by value [-Werror=catch-value=]
 3519 |   } catch(std::range_error e) {
      |                            ^
cc1plus: all warnings being treated as errors
ninja: build stopped: subcommand failed.
[developer/src/kmcmplib] ## build:arch failed
```

### PR DESCRIPTION
**Pull request by markcsinclair**
Attempt to get the kmcmplib compiler (and the new google tests) running under linux - fixes #11772

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by markcsinclair** [comment]
It is now passing full-test on Windows, but I will recheck the changes for correctness.

**Comment by markcsinclair** [comment]
A minimal set of changes to get kmcmplib compiling under linux.  An error (unterminated strings) in ValidateMatchNomatchOutput_test was detected and corrected; this will also be corrected in PR [#11663](https://github.com/keymanapp/keyman/pull/11663).

Ready for review.

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 1]
```diff
@@ -3516,7 +3517,7 @@ bool UTF16TempFromUTF8(KMX_BYTE* infile, int sz, KMX_BYTE** tempfile, int *sz16)
   try {
     std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> converter;
     result = converter.from_bytes((char*)infile, (char*)infile+sz);
-  } catch(std::range_error e) {
+  } catch(std::range_error &e) {
```
Comment (ermshiperete) [inline developer/src/kmcmplib/src/Compiler.cpp:3520]: Not that it matters, but I find this easier to read:
```suggestion
  } catch(std::range_error& e) {
```

File: developer/src/kmcmplib/src/kmx_u16.cpp [hunk 2]
```diff
@@ -10,6 +10,11 @@
 #include <locale>
 #include <stdarg.h>
 
+#ifdef _MSC_VER
+#else
+#include <algorithm>
+#endif
```
Comment (ermshiperete) [inline developer/src/kmcmplib/src/kmx_u16.cpp:16]: maybe?
```suggestion
#ifndef _MSC_VER
#include <algorithm>
#endif
```

File: developer/src/kmcmplib/src/kmx_u16.cpp [hunk 3]
```diff
@@ -96,7 +101,11 @@ std::string toHex(int num1) {
 			s += (87 + temp);
 		num = num / 16;
 	}
+#ifdef _MSC_VER
 	reverse(s.begin(), s.end());
+#else
+	std::reverse(s.begin(), s.end());
```
Comment (ermshiperete) [inline developer/src/kmcmplib/src/kmx_u16.cpp:107]: I wonder if `std::reverse` would also work on Windows.

File: developer/src/kmcmplib/tests/kmcompxtest.cpp [hunk 4]
```diff
@@ -102,7 +104,7 @@ int main(int argc, char *argv[])
       std::istringstream(ErrNr) >> std::hex >> error_val;
 
       // check if error_val is in Array of Errors; if it is found return 0 (it's not an error)
-      for (int i = 0; i < error_vec.size() ; i++) {
+      for (std::vector<int>::size_type i = 0; i < error_vec.size() ; i++) {
```
Comment (ermshiperete) [inline developer/src/kmcmplib/tests/kmcompxtest.cpp:107]: Would `size_t` work? That would be a bit simpler.
```suggestion
      for (size_t i = 0; i < error_vec.size() ; i++) {
```

**Review by ermshiperete — APPROVED** [review]
A few comments and suggestions, but overall LGTM.

**Review by mcdurdin — APPROVED** [review]

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 1 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/src/Compiler.cpp:3520]: Okay.

File: developer/src/kmcmplib/src/kmx_u16.cpp (hunk 3 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/src/kmx_u16.cpp:107]: It does seem to work ... I had adopted the more conservative change, but I'll switch to std::reverse for all.

File: developer/src/kmcmplib/tests/kmcompxtest.cpp (hunk 4 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/kmcompxtest.cpp]: Seems to work. Done.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.57-alpha](https://keyman.com/downloads/releases/alpha/18.0.57)

**Pull request merged**